### PR TITLE
[TT-Transformers] Add padding to vocab-size to improve top-k performance on multichip systems

### PR DESCRIPTION
**Pull request by ifabijanic-tt**
### Problem description
Llama-3.3-70B, Qwen-3-32B and other models had a performance drop in throughput. For Llama-3.3-70B it went from 15.2 tok/s/user to 14 tok/s/user on T3K devices on the `performance-ci-32 test`. 

### What's changed
The perf drop happened with this commit: https://github.com/tenstorrent/tt-metal/commit/927a5409f43b370e00053805fe2d5206ed87f4c0

This commit included changes to top-k calculations and fixes on this op. One of the changes was the inclusion of bionic sort which required inputs to be in power of 2 sizes to work correctly. Because of that additional check was added which checked for that. If the check failed decode sampling of Top is not done on multiple chips but failed back to single core calculation. This would decrease performance significantly on multi chip devices. 

Llama-3.3-70B has vocabulary size of 128256 and in the current code `compute_padded_vocab_size()` only pads to a tile-aligned multiple of `32 * num_devices`, not a power of two, in `model_config.py`. On a Loudbox -> TP=8, the per-device width is `128256 / 8 = 16032`, which passes the old multi-core checks (`>=8192, <65536, k<=64`). However this number is not a power of 2 so it fails the new power-of-two check. Because of that the decode sampling of TopK falls back to single core and this kill throughput.

The fix in this PR checks enforces a check on TT-Transformer models to see if the vocab_size // num_devices is a power of two, and pads it if needed.
For Llama-3.3-70B, this fix increases performance from previous 14 tok/s/user to 16.2 tok/s/user.

#### Model tests
- [ ] T3K Demo - https://github.com/tenstorrent/tt-metal/actions/runs/23487549333
- [ ] Single card demo - https://github.com/tenstorrent/tt-metal/actions/runs/23487554651

### CODE REVIEW

File: models/tt_transformers/tt/model_config.py [hunk 1]
```diff
@@ -88,6 +88,19 @@ def compute_padded_vocab_size(vocab_size: int, num_devices: int) -> int:
     return nearest_multiple(vocab_size, ttnn.TILE_SIZE * num_devices)
 
 
+def should_pad_sampling_logits_to_power_of_2(base_model_name: str, padded_vocab_size: int, sampling_splits: int) -> bool:
+    # Enable optional sampling padding for models that regress to single-core TopK.
+    if sampling_splits < 1:
+        raise ValueError(f"sampling_splits must be >= 1, got {sampling_splits}")
```
Comment (mtairum) [inline models/tt_transformers/tt/model_config.py]: raising an error might be overkill, no? Maybe a warning, as we already do the hardcode below?

File: models/tt_transformers/tt/model_config.py [hunk 2]
```diff
@@ -88,6 +88,19 @@ def compute_padded_vocab_size(vocab_size: int, num_devices: int) -> int:
...
+
+    # This is hardcoded on purpose just to target Llama-3.1-70B for the fix. If removed then all models receive a fix
+    if base_model_name != "Llama-3.1-70B":
```
Comment (mtairum) [inline models/tt_transformers/tt/model_config.py]: Mention here the issue you've created, for reference. also mention that this is Llama only since that's where we measured regression. More models might be added later.

And this needs to be for llama-3.1-70B and llama-3.3-70B (since we're moving for the more recent weights). vocab size is the same in both 👍

File: models/tt_transformers/tt/model_config.py [hunk 3]
```diff
@@ -2614,6 +2627,17 @@ def _set_params_from_dict(self, config):
                 "Qwen2.5-7B and Qwen2.5-VL-7B is only supported on 2 or 4 devices, run on an N300 or use MESH_DEVICE=N150x4"
             )
 
+        if self.num_devices:
```
Comment (mtairum) [inline models/tt_transformers/tt/model_config.py]: isn't self.num_devices always defined? 

And this pad is required for every use-case where we split the vocab size right? Sometimes we do split the vocab when devices==1, since we have to split computations to fit in memory

**Review by mtairum — APPROVED** [review]
Pre-approving to unblock.

Please address the comments before merge🙏

**Review by gwangTT — APPROVED** [review]

**Review by mtairum — APPROVED** [review]
Re-reviewed 👍
